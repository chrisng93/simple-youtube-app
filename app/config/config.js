export const YOUTUBE_API_KEY = 'AIzaSyAz_sPxKJzgqgnIhdR8vonCSJHOFBV2VpY';
