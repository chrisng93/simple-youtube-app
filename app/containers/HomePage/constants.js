export const SEARCH_VIDEOS = 'SEARCH_VIDEOS';
export const SEARCH_VIDEOS_SUCCESS = 'SEARCH_VIDEOS_SUCCESS';
export const SEARCH_VIDEOS_FAILURE = 'SEARCH_VIDEOS_FAILURE';

export const SELECT_VIDEO = 'SELECT_VIDEO';
