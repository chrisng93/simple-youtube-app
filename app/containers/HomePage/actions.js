import { SEARCH_VIDEOS, SEARCH_VIDEOS_SUCCESS, SEARCH_VIDEOS_FAILURE, SELECT_VIDEO } from './constants';

export const searchVideos = (payload) => {
  return {
    type: SEARCH_VIDEOS,
    payload,
  }
};

export const searchVideosSuccess = (payload) => {
  return {
    type: SEARCH_VIDEOS_SUCCESS,
    payload,
  }
};

export const searchVideosFailure = (payload) => {
  return {
    type: SEARCH_VIDEOS_FAILURE,
    payload,
  }
};

export const selectVideo = (payload) => {
  return {
    type: SELECT_VIDEO,
    payload,
  }
};
