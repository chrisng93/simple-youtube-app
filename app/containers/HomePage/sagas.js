import { take, call, put, select, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import request from '../../utils/request';
import { searchVideosSuccess, searchVideosFailure } from './actions';
import { YOUTUBE_API_KEY } from '../../config/config';
import { SEARCH_VIDEOS } from './constants';

export function* getVideos({ payload: { query } }) {
  const queryString = `part=snippet&type=embeddable&video=true&maxResults=20&key=${YOUTUBE_API_KEY}&q=${query}`;
  const requestUrl = `https://www.googleapis.com/youtube/v3/search?${queryString}`;

  try {
    const response = yield call(request, requestUrl);
    yield put(searchVideosSuccess({ videos: response.items }));
  } catch(error) {
    yield put(searchVideosFailure({ message: error }));
  }
}

export function* getVideosWatcher() {
  const watcher = yield takeLatest(SEARCH_VIDEOS, getVideos);

  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

export default [
  getVideosWatcher,
]
