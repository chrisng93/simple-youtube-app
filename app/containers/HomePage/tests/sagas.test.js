import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';
import { LOCATION_CHANGE } from 'react-router-redux';
import { YOUTUBE_API_KEY } from '../../../config/config';
import { SEARCH_VIDEOS } from '../constants';
import { searchVideosSuccess, searchVideosFailure } from '../actions';
import { getVideos, getVideosWatcher } from '../sagas';

const query = 'test';

/* eslint-disable redux-saga/yield-effects */
describe('getVideos Saga', () => {
  let getVideosGenerator;

  beforeEach(() => {
    getVideosGenerator = getVideos({ payload: { query } });

    const callDescriptor = getVideosGenerator.next(YOUTUBE_API_KEY, query).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the searchVideosSuccess action if it requests the data successfully', () => {
    const response = {
      items: [{ test: 'test' }],
    };
    const putDescriptor = getVideosGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(searchVideosSuccess({ videos: response.items })));
  });

  it('should dispatch the searchVideosFailure action if the response errors', () => {
    const response = new Error('some error');
    const putDescriptor = getVideosGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(searchVideosFailure({ message: response })));
  });
});

describe('getVideosWatcher', () => {
  const getVideosWatcherSaga = getVideosWatcher();
  const mockedTask = createMockTask();

  it('should start task to watch for SEARCH_VIDEOS action', () => {
    const takeLatestDescriptor = getVideosWatcherSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(SEARCH_VIDEOS, getVideos));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    const takeDescriptor = getVideosWatcherSaga.next(mockedTask).value;
    expect(takeDescriptor).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel the forked task when LOCATION_CHANGE happens', () => {
    const cancelDescriptor = getVideosWatcherSaga.next().value;
    expect(cancelDescriptor).toEqual(cancel(mockedTask));
  });
});
