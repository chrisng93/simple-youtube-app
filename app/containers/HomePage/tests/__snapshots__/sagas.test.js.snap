exports[`getVideos Saga should dispatch the searchVideosFailure action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://www.googleapis.com/youtube/v3/search?part=snippet&type=embeddable&video=true&maxResults=20&key=AIzaSyAz_sPxKJzgqgnIhdR8vonCSJHOFBV2VpY&q=test",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getVideos Saga should dispatch the searchVideosSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://www.googleapis.com/youtube/v3/search?part=snippet&type=embeddable&video=true&maxResults=20&key=AIzaSyAz_sPxKJzgqgnIhdR8vonCSJHOFBV2VpY&q=test",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
